src/main/java/org/testng/JarFileUtils.java

```diff
@@ -52,23 +59,41 @@ private boolean testngXmlExistsInJar(File jarFile, List<String> classes) throws
         boolean foundTestngXml = false;
         try (JarFile jf = new JarFile(jarFile)) {
             Enumeration<JarEntry> entries = jf.entries();
+            File file = java.nio.file.Files.createTempDirectory("testngXmlPathInJar-").toFile();
+            String suitePath = null;
             while (entries.hasMoreElements()) {
                 JarEntry je = entries.nextElement();
-                if (matchesXmlPathInJar(je)) {
-                    Collection<XmlSuite> parsedSuites = Parser.parse(jf.getInputStream(je), processor);
-                    for (XmlSuite suite : parsedSuites) {
-                        // If test names were specified, only run these test names
-                        XmlSuite suiteToAdd = suite;
-                        if (testNames != null) {
-                            suiteToAdd = XmlSuiteUtils.cloneIfContainsTestsWithNamesMatchingAny(suite, testNames);
-                        }
-                        suites.add(suiteToAdd);
+                String jeName = je.getName();
+                if (Parser.canParse(jeName.toLowerCase())){ 
+                    InputStream  inputStream = jf.getInputStream(je);
+                    File copyFile = new File(file, jeName);
+                    Files.copyFile(inputStream, copyFile);
+                    if (matchesXmlPathInJar(je)) {
+                        suitePath = copyFile.toString();
                     }
-                    return true;
                 } else if (isJavaClass(je)) {
                     classes.add(constructClassName(je));
                 }
             }
+            if (Strings.isNullOrEmpty(suitePath)) {
+                return foundTestngXml;
+            }
+            Collection<XmlSuite> parsedSuites = Parser.parse(suitePath,processor); 
+            for (XmlSuite suite : parsedSuites) {
+                // If test names were specified, only run these test names
+                if (testNames != null) {
```

Comment:
What if `testNames` is empty? Instead of checking against null, we should not accept null as value because an empty list is easy to pass.

Reply:
`if (testNames != null) {` this logic has existed long before my commit. But I think here it is correct. 
1. Default value is null, there is no chance to change it to empty, except misuse -testNames
2. Even misuse, "cloneIfContainsTestsWithNamesMatchingAny" would check if it is empty.

Reply:
In that case, translate the null value by an `Collections.emptyList()` in the constructor when the param is null. It will simplify the logic here.

Reply:
I don't think so. Default value of `List<String> testNames` setting as null is meanful, which means `-testNames` is not active.

Reply:
What is the different behavior between null and empty?

Reply:
1. null means function ` -testNames` is not active as default.
2. empty means user miss-used ` -testNames` where an exception should be exposed as expected, like what it does right now:
'throw new TestNGException("Please provide a valid list of names to check.");'